test/framework/machinehealthcheck_helpers.go

```diff
@@ -0,0 +1,169 @@
+		Status:             mhc.Spec.UnhealthyConditions[0].Status,
+		LastTransitionTime: metav1.Time{Time: time.Now()},
+	}
+	for _, md := range input.MachineDeployments {
+		selectorMap, err := metav1.LabelSelectorAsMap(&md.Spec.Selector)
+		Expect(err).ToNot(HaveOccurred())
+
+		ms := &clusterv1.MachineSetList{}
+		Expect(input.ClusterProxy.GetClient().List(ctx, ms, client.InNamespace(input.Cluster.Namespace), client.MatchingLabels(selectorMap))).To(Succeed())
+		Expect(len(ms.Items)).NotTo(Equal(0))
+
+		machineSet := ms.Items[0]
+		selectorMap, err = metav1.LabelSelectorAsMap(&machineSet.Spec.Selector)
+		Expect(err).ToNot(HaveOccurred())
+		machines := &clusterv1.MachineList{}
+		Expect(input.ClusterProxy.GetClient().List(ctx, machines, client.InNamespace(machineSet.Namespace), client.MatchingLabels(selectorMap))).To(Succeed())
+
+		for _, machine := range machines.Items {
+			node := &corev1.Node{}
+			Expect(input.ClusterProxy.GetWorkloadCluster(ctx, input.Cluster.Namespace, input.Cluster.Name).GetClient().Get(ctx, types.NamespacedName{Name: machine.Status.NodeRef.Name, Namespace: machine.Status.NodeRef.Namespace}, node)).To(Succeed())
+			patchHelper, err := patch.NewHelper(node, input.ClusterProxy.GetWorkloadCluster(ctx, input.Cluster.Namespace, input.Cluster.Name).GetClient())
+			Expect(err).ToNot(HaveOccurred())
+			node.Status.Conditions = append(node.Status.Conditions, unhealthyNodeCondition)
+			Expect(patchHelper.Patch(ctx, node)).To(Succeed())
+		}
+	}
+
+	fmt.Fprintf(GinkgoWriter, "Patching MachineHealthCheck to initiate reconcile\n")
+	Expect(input.ClusterProxy.GetClient().Get(ctx, client.ObjectKey{Namespace: input.Cluster.Namespace, Name: input.MachineHealthCheckName}, mhc)).To(Succeed())
+	patchHelper, err := patch.NewHelper(mhc, input.ClusterProxy.GetClient())
+	Expect(err).ToNot(HaveOccurred())
+	mhc.Labels["e2e.x-k8s.io"] = "triggered"
+	Expect(patchHelper.Patch(ctx, mhc)).To(Succeed())
+
+	fmt.Fprintf(GinkgoWriter, "Waiting for MachineHealthCheck to detect unhealthy nodes\n")
+	Eventually(func() bool {
+		Expect(input.ClusterProxy.GetClient().Get(ctx, client.ObjectKey{Namespace: input.Cluster.Namespace, Name: input.MachineHealthCheckName}, mhc)).To(Succeed())
+		return mhc.Status.ExpectedMachines > 0 && mhc.Status.CurrentHealthy != mhc.Status.ExpectedMachines
+	}, intervals...).Should(BeTrue())
+
+	fmt.Fprintf(GinkgoWriter, "Waiting for MachineHealthCheck to remediate unhealthy nodes\n")
+	counter := 0
+	Eventually(func() bool {
+		// Patching MachineHealthCheck here to avoid waiting for MachineHealthCheck to detect the changes
+		Expect(input.ClusterProxy.GetClient().Get(ctx, client.ObjectKey{Namespace: input.Cluster.Namespace, Name: input.MachineHealthCheckName}, mhc)).To(Succeed())
+		patchHelper, err := patch.NewHelper(mhc, input.ClusterProxy.GetClient())
+		Expect(err).ToNot(HaveOccurred())
+		mhc.Labels["e2e.x-k8s.io"] = "triggered" + strconv.Itoa(counter)
+		Expect(patchHelper.Patch(ctx, mhc)).To(Succeed())
```

Comment:
It's a little strange that we have to patch the MHC to get it to remediate, these tests are effectively tapping in the controller's internal implementations. When soon #2975 gets merged for example, this logic wouldn't be correct anymore and we'd have to change code in two places